src/main/kotlin/dev/sparky200/skriptnbt/nbt/NbtCompound.kt

```diff
@@ -0,0 +1,95 @@
+package dev.sparky200.skriptnbt.nbt
+
+import org.jetbrains.annotations.ApiStatus.Experimental
+
+@Experimental
+class NbtCompound(val resolver: NbtTagResolver = NbtTagResolver.Default) : NbtTag<Map<String, Any?>> {
+    private val map = mutableMapOf<String, NbtTag<*>>()
+
+    override val value: Map<String, Any?>
+        get() = map.entries.associate { (key, value) -> key to value.value }
+
+    val keys: Collection<String>
+        get() = map.keys
+
+    // TODO: type it!
+    operator fun set(key: String, value: Any?) {
+        map[key] = resolver.resolve(value)
+    }
+
+    fun setTag(key: String, tag: NbtTag<*>) {
+        map[key] = tag
+    }
+
+    operator fun get(key: String): Any? = map[key]?.value
+    fun getAsTag(key: String): NbtTag<*>? = map[key]
+
+    inline fun <reified T> getAs(key: String): T? {
+        val tag = getAsTag(key)
+        if (tag is T) return tag
+        if (tag != null && tag.value is T) return tag.value as T
+        return null
+    }
+
+    fun clear() {
+        map.clear()
+    }
+
+    fun remove(key: String): Any? {
+        return map.remove(key)?.value
+    }
+
+    operator fun contains(key: String): Boolean = key in map
+
+    override fun toNbtString(): String = map.entries.joinToString(separator = ",", prefix = "{", postfix = "}") { (key, value) -> "$key=${value.toNbtString()}" }
+
+    override fun toString(): String {
+        return "NbtCompound{${map.entries.joinToString { "\"${it.key}\"=\"${it.value}\"" }}}"
+    }
+}
+
+fun NbtCompound(resolver: NbtTagResolver, vararg elements: Pair<String, Any?>): NbtCompound {
+    val compound = NbtCompound(resolver)
+    elements.forEach { (key, value) ->
+        compound[key] = value
+    }
+    return compound
+}
+
+fun NbtCompound(vararg elements: Pair<String, Any?>): NbtCompound = NbtCompound(NbtTagResolver.Default, *elements)
+
+private fun parseArrayOrListTag(value: String): NbtTag<*> {
+    TODO()
+}
+
+private fun parseStringTag(value: String): NbtTag<*> {
+    if (value.startsWith("\"")) {
```

Comment:
So you need to be _very_ careful here: for legacy reasons, strings in snbt inputs **don't** require quotes as long as they meet these criteria:
> A string enclosed in quotes. For strings containing only 0-9, A-Z, a-z, _, -, ., and +, and not confused with other data types, quote enclosure is optional.

You probably want a fall-through case that checks the value against this regex, but you'll need to be careful of the number postfix options (e.g. something ending `s` could be a short value like `10s` or it could be a non-quoted simplestring like `sausages`)

There's more information on the wiki here <https://minecraft.wiki/w/NBT_format#SNBT_format> :)

Reply:
Of course they don't! Making things harder to parse is always better!